resolver/src/resolver_handle/background.rs

```diff
@@ -0,0 +1,224 @@
+use std::sync::{Arc, Mutex};
+
+use futures::{
+    future,
+    sync::mpsc,
+    Async, Future, Poll, Stream
+};
+use trust_dns_proto::{
+    error::ProtoResult,
+    rr::{Name, RData, RecordType},
+    xfer::{ DnsRequestOptions, RetryDnsHandle,},
+};
+#[cfg(feature = "dnssec")]
+use trust_dns_proto::SecureDnsHandle;
+
+use config::{ResolverConfig, ResolverOpts};
+use dns_lru::DnsLru;
+use hosts::Hosts;
+use lookup::{Lookup, LookupEither, LookupFuture};
+use lookup_ip::LookupIpFuture;
+use lookup_state::CachingClient;
+use name_server_pool::{NameServerPool, StandardConnection};
+
+use super::{BasicResolverHandle, Request};
+
+/// Returns a new future that will drive the background task.
+pub(super) fn task(
+    config: ResolverConfig,
+    options: ResolverOpts,
+    lru: Arc<Mutex<DnsLru>>,
+    request_rx: mpsc::UnboundedReceiver<Request>,
+) -> impl Future<Item = (), Error = ()> {
+    future::lazy(move || {
+        let pool = NameServerPool::<BasicResolverHandle, StandardConnection>::from_config(
+            &config, &options,
+        );
+        let either;
+        let client = RetryDnsHandle::new(pool.clone(), options.attempts);
+        if options.validate {
+            #[cfg(feature = "dnssec")]
+            {
+                either = LookupEither::Secure(SecureDnsHandle::new(client));
+            }
+
+            #[cfg(not(feature = "dnssec"))]
+            {
+                // TODO: should this just be a panic, or a pinned error?
+                warn!("validate option is only available with 'dnssec' feature");
+                either = LookupEither::Retry(client);
+            }
+        } else {
+            either = LookupEither::Retry(client);
+        }
+
+        let hosts = if options.use_hosts_file {
+            Some(Arc::new(Hosts::new()))
+        } else {
+            None
+        };
+
+        Task {
+            config,
+            options,
+            client_cache: CachingClient::with_cache(lru, either),
+            hosts: hosts,
+            request_rx,
+        }
+    })
+}
+
+type ClientCache =
+    CachingClient<LookupEither<BasicResolverHandle, StandardConnection>>;
+
+/// Background task that resolves DNS queries.
+struct Task {
+    config: ResolverConfig,
+    options: ResolverOpts,
+    client_cache: ClientCache,
+    hosts: Option<Arc<Hosts>>,
+    request_rx: mpsc::UnboundedReceiver<Request>,
+}
+
+impl Task {
+
+    fn lookup(
+        &self,
+        name: Name,
+        record_type: RecordType,
+        options: DnsRequestOptions,
+    ) -> LookupFuture {
+        let names = self.build_names(name);
+        LookupFuture::lookup(names, record_type, options, self.client_cache.clone())
+    }
+
+    fn lookup_ip(
+        &self,
+        maybe_name: ProtoResult<Name>,
+        maybe_ip: Option<RData>
+    ) -> LookupIpFuture {
+        let mut finally_ip_addr = None;
+
+        // if host is a ip address, return directly.
+        if let Some(ip_addr) = maybe_ip {
```

Comment:
I think we should move this logic out to the initial lookup call, so that we can remove that 'static. This doesn't need to be part of the future IMO... The cases of the future being canceled before this logic is run will probably be rare, and therefor the cost sames don't seem worth the 'static restriction on the initial call.

EDIT: I made this commend before realizing that the Handler's lookup_ip() doesn't actually have 'static on it...